controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -399,45 +420,62 @@ public String getScopeName() {
     }
 
     @Override
-    public CompletableFuture<TxnStatus> sealTransaction(final UUID txId, final boolean commit,
-                                                        final Optional<Integer> version) {
-        return verifyLegalState(checkTransactionStatus(txId)
-                .thenCompose(x -> {
+    public CompletableFuture<SimpleEntry<TxnStatus, Integer>> sealTransaction(final UUID txId,
+                                                                              final boolean commit,
+                                                                              final Optional<Integer> version) {
+        CompletableFuture<SimpleEntry<TxnStatus, Integer>> future = verifyLegalState(getTransactionEpoch(txId)
+                .thenCompose(epoch -> sealActiveTxn(epoch, txId, commit, version)))
+                .exceptionally(ex -> new SimpleEntry<>(handleDataNotFoundException(ex), null));
```

Comment:
The handles is expected to intercept and act only on DataNotFoundException; other exceptions are rethrown. Possibly, I could rename the method to `handleError`.